---
title: "Overview"
teaching: 20
exercises: 0
questions:
objectives:
- "Topics and technologies used during this workshop"
keypoints:
- "This part of the lesson will mostly be given in the form of a lecture"
apps:
- "Powerpoint"
---

## Target audience and main aim

This two-day workshop is aimed at providing a basic understanding of creating and evaluating de novo assembly using long read technologies. The workshop requires knowledge of working with the Linux operating system and of biological topics such as genetics and DNA.
The target audience consists of Master/PhD students or researchers working in the Life Sciences. Although the data is from a bacterial genome, the learned principles can be applied to many other organisms, including fungi, plants and mammals. 
The idea behind this workshop is that you learn the most by doing and discussing the results.    

## Topics

1. **Read Quality assessment.** Using bioinformatics tools to check the quality of the reads.
2. **De novo assembly.**  Producing genome assemblies based on PacBio and Nanopore.
3. **Assembly comparison.** The genome assemblies based on the different types of data might have different results. Here we will compare the results and investigate the found differences between them. 
4. **Assembly base quality** We will use our NGS data to identify SNPs and other discrepancies between the assemblies. We will study the effects these assembly errors have on downstream analyzes.

## Google drive

To give feedback, solutions and other things you'd like to share, please use this [Google document](https://docs.google.com/document/d/10Vii37lR8TYYvJnYartqw2RxHooBfiWagJyHWOd6qXs/edit?usp=sharing).

## Presentation

[The presentation on long read technologies is also available for download.](../fig/LongReadAssembly.pptx)

{% include links.md %}
