---
title: "QC and evaluation of long read data"
teaching: 30
exercises: 60-90
questions:
- "How to determine basic statistics on read sets"
- "What do these statistics mean: are they good or bad?"
- "What are read technology specific errors?"  
objectives:
- "Learn how to get and interpret read set statistics"
- "Evaluate read technology specific errors based on read mappings"
keypoints:
- "PacBio and Nanopore have read lengths of many kilobases"
- "In practice quality of input DNA, library prep. and many other factors will determine read length distribution"
- "PacBio and Nanopore have different types of errors"
apps:
- "assembly-stats" 
- "minimap2"
- "./tools/Tablet/tablet"
---

## Introduction

Quality control of input data is an important task in any data analysis and is very much true for de novo assemblies. Biases, high error reads and other unexpected properties will have an impact on the resulting contigs by producing fragmented assemblies and/or chimeric contigs. In this session we will have a closer look at our input data before we continue with the assembly process.

> ## Reads
> The data we are using for this workshop are reads selected from a ~2.5M region in the kiwifruit genome. Assembling a full genome will take too long and a laptop is not a suitable device to assembly such a large genome. There are four read files:
~~~
data/reads/hifi_reads.fastq
data/reads/hifi_sub_reads.fastq
data/reads/ont_reads.fastq
data/reads/ont_sub_reads.fastq
~~~
{: bash}

## Read statistics
For basic the quality control of the read sets we use assembly-stats:
~~~
{{site.vm_prompt}}assembly-stats 
usage: stats [options] <list of fasta/q files>

Reports sequence length statistics from fasta and/or fastq files

options:
-l <int>
	Minimum length cutoff for each sequence.
	Sequences shorter than the cutoff will be ignored [1]
-s
	Print 'grep friendly' output
-t
	Print tab-delimited output
-u
	Print tab-delimited output with no header line
-v
	Print version and exit
~~~
{: .bash}

This tool will report to the terminal information like average sequence length and N50. 

> ## Read statistics
> The PacBio reads are in this file: data/reads/hifi_reads.fastq
> Get the read statistics on this data set and try to interpret the results.
> > ## Solution
> > ~~~
> > {{site.vm_prompt}}assembly-stats data/reads/hifi_reads.fastq 
> > stats for hifi_reads.fastq
> > sum = 1130395093, n = 128825, ave = 8774.66, largest = 50239
> > N50 = 14387, n = 32641
> > N60 = 13679, n = 40702
> > N70 = 13026, n = 49173
> > N80 = 12416, n = 58064
> > N90 = 11792, n = 67398
> > N100 = 200, n = 128825
> > N_count = 0
> > Gaps = 0
> > ~~~
> > {: .bash}
> {: .solution}
> The Nanopore reads are in this file: data/reads/ont_reads.fastq. Also check the statistics of the Nanopore data.
> > ## Solution
> > ~~~
> > {{site.vm_prompt}}assembly-stats data/reads/ont_reads.fastq 
> > stats for ont_reads.fastq
> > sum = 604909161, n = 90323, ave = 6697.18, largest = 439418
> > N50 = 87997, n = 2515
> > N60 = 78457, n = 3245
> > N70 = 69903, n = 4061
> > N80 = 61085, n = 4987
> > N90 = 5665, n = 6652
> > N100 = 42, n = 90323
> > N_count = 0
> > Gaps = 0
> > ~~~
> > {: .bash}
> {: .solution}
> Please interpret the results of the two output. Discuss the difference between the two platforms, if there are any.
> There is something strange about the HiFi statistics. Do you spot this?
{: .challenge}

## Aligning the reads to the reference genome

A further visual inspection of the data can be done by mapping the reads to the reference genome. For each platform many tools exist to help you with this. We will focus on the newest and/or most commonly used tools.
The alignments will give you ideas on SNP rates, biases in the data, heterozygosity and error rates.

> ## The reference genome 
> The reads are from the kiwifruit. We can expect that the reads have high similarity when compare to this genome. However, some issues have to be taken into account when mapping reads to a genome. Discuss what effect will be of the following issues when looking at the mapped reads:
> 
> 1. The genome of the organism is not known. You will have to map the reads to the genome of a relative.  
> 2. Your reference genome is highly repetitive.
> 3. Your reference genome is not nearly complete.
> 4. You have a polyploid organism.
> 5. You are faced with issues 2-5
{: .challenge}

For mapping the reads we will use [minimap2](https://github.com/lh3/minimap2/blob/master/README.md):
~~~
minimap2
~~~
{: .bash}

And a 2.5MB region of the kiwifruit genome:
~~~
data/genome/kiwi_contig.fasta
~~~
{: .bash}
 
> ## Mapping reads
> Mapping reads is a memory and compute intensive task. When limited resources are available, please use the subsampled read set for the mappings.
>
> Use minimap2 to map the two read sets to our region. The PacBio data:
> ~~~
> data/reads/hifi_reads.fastq
> ~~~
> {: .bash}
> The nanopore data:
> ~~~
> data/reads/ont_reads.fastq
> ~~~
> {: .bash}
> Run minimap2 without arguments to see which options you can use. For subsequent steps you need SAM output. 
> > ## Solution
> > ~~~
> > {{site.vm_prompt}}mkdir results
> > {{site.vm_prompt}}minimap2 -a -t 6 -x map-hifi --secondary=no ./data/genome/kiwi_contig.fa ./data/reads/hifi_reads.fastq > results/ref_hifi.sam
> > {{site.vm_prompt}}minimap2 -a -t 6 -x map-ont --secondary=no ./data/genome/kiwi_contig.fa ./data/reads/ont_reads.fastq > results/ref_ont.sam
> > ~~~
> > {: .bash}
> {: .solution}
{: .challenge}

> ## Visualizing the results
> Tablet is a nice and compact SAM/BAM viewer. It is written in Java and it runs on most systems.
> ~~~
> {{site.vm_prompt}}~/tools/Tablet/tablet
> ~~~
> {: .bash}
> Open Tablet and load the reference plus a SAM file. For each of the three results files, visually check:
> 
> 1. The number of mismatches, insertions and deletions (Tablet supports different **Color schemes** to help with this).
> 2. Coverage across the region
> 
> Discuss the differences between the two results. For example:
> 
> 1. What type of errors do you see? 
> 2. How many of them?
> 3. What will be the likely impact on the assembly?
>
> ## Large miss-alignments
> The alignment file contains many reads that show high inaccuracy. What could be a reason for this?
> > ## Solution
> > In our case we map reads from a genomic sample to only a part of the genome. Genomes contain many parts which are (very) similar, like gene copies, repeat regions, etc.
> > Reads which originate from these parts of the genome will map to our smaller region, usually with much less accuracy. 
> > When available, you should always use the whole genome to map reads from a genomic samples. Preferably including mtDNA and chloroplast.
> > Be aware, that only a very few genomes are available 'telomere to telomere', and usually you are forced to use a closely related species (also for bacteria and fungi) with an incomplete genome. Always be on the look-out for these false-positive mapping results!
> > In this case we can remove most false positives by selecting reads with hardly any overlapping nucleotides (soft clippings):
> > ~~~
> > {{site.vm_prompt}}grep -v "[0-9]\{2,\}S" ref_hifi.sam > ref_hifi_filtered.sam
> > ~~~
> > {: .bash}
> {: .solution}
{: .challenge}

> ## Kiwifruit reference genome
> Our cultivar is a diploid plant with two clear haplotypes.
> Discuss based on the mapping results:
> 
> 1. Most obvious differences between the mapping results.
> 2. The reference sequences is a collapsed haplotype. What is the impact of this on the mapping results? 
> 3. Do you see any differences in mapping information for each of the sequencing platforms? 
> 4. After assembly, with what be key differences between the assemblied sequences and the kiwifruit genome?
> 5. Why would you need either HiFi PacBio / High quality base calling Nanopore and not (much longer) PacBio subreads / basic quality base calling Nanopore?
> 6. Kiwi has a diploid genome. What are issues that you might encounter when sequencing and assembling a metagenomics sample?
{: .challenge}
